File: price_scraper.js
import fs from 'fs';
import csv from 'csv-parser';
import { createObjectCsvWriter } from 'csv-writer';
import { chromium } from 'playwright';

const INPUT_CSV = 'products_export.csv';
const OUTPUT_CSV = 'products_with_prices.csv';

// 读取商品标题
function readProductTitles(inputFile) {
  return new Promise((resolve, reject) => {
    const titles = [];
    fs.createReadStream(inputFile)
      .pipe(csv())
      .on('data', (row) => {
        if (row['Title'] && !titles.includes(row['Title'])) {
          titles.push(row['Title']);
        }
      })
      .on('end', () => {
        resolve(titles);
      })
      .on('error', reject);
  });
}

// 读取所有行，后续写回
function readAllRows(inputFile) {
  return new Promise((resolve, reject) => {
    const rows = [];
    fs.createReadStream(inputFile)
      .pipe(csv())
      .on('data', (row) => rows.push(row))
      .on('end', () => resolve(rows))
      .on('error', reject);
  });
}

async function searchAmazon(page, title) {
  try {
    await page.goto('https://www.amazon.com', { timeout: 60000 });
    await page.fill('#twotabsearchtextbox', title);
    await page.keyboard.press('Enter');
    await page.waitForSelector('[data-component-type="s-search-result"]', { timeout: 60000 });
    const price = await page.locator('.a-price .a-offscreen').first().textContent();
    // 点击第一个商品，进入详情页
    await page.locator('[data-component-type="s-search-result"]').first().click();
    await page.waitForLoadState('domcontentloaded', { timeout: 60000 });
    // 抓取描述（例如，在 #productDescription 或 #feature-bullets 中）
    const description = await page.locator('#productDescription, #feature-bullets').textContent().catch(() => '');
    // 抓取商品尺寸（例如，在 #detailBullets_feature_div 中，或通过正则匹配）
    const sizeText = await page.locator('#detailBullets_feature_div, #prodDetails').textContent().catch(() => '');
    const sizeMatch = sizeText.match(/size\s*:?\s*([^,;]+)/i);
    const productSize = sizeMatch ? sizeMatch[1].trim() : '';
    console.log("Amazon 抓取字段:", { price, description, productSize });
    return { price: price ? price.trim() : '', description, productSize };
  } catch (e) {
    console.log("Amazon 抓取错误:", e);
    return { price: '', description: '', productSize: '' };
  }
}

async function searchEbay(page, title) {
  try {
    await page.goto('https://www.ebay.com', { timeout: 60000 });
    await page.fill('#gh-ac', title);
    await page.keyboard.press('Enter');
    await page.waitForSelector('.s-item__price', { timeout: 60000 });
    const price = await page.locator('.s-item__price').first().textContent();
    // 点击第一个商品，进入详情页
    await page.locator('.s-item__link').first().click();
    await page.waitForLoadState('domcontentloaded', { timeout: 60000 });
    // 抓取描述（例如，在 #itemDescription 中）
    const description = await page.locator('#itemDescription').textContent().catch(() => '');
    // 抓取商品尺寸（例如，在 .ux-labels-values__labels 中，或通过正则匹配）
    const sizeText = await page.locator('.ux-labels-values__labels, .ux-labels-values__values').textContent().catch(() => '');
    const sizeMatch = sizeText.match(/size\s*:?\s*([^,;]+)/i);
    const productSize = sizeMatch ? sizeMatch[1].trim() : '';
    console.log("eBay 抓取字段:", { price, description, productSize });
    return { price: price ? price.trim() : '', description, productSize };
  } catch (e) {
    console.log("eBay 抓取错误:", e);
    return { price: '', description: '', productSize: '' };
  }
}

async function searchWalmart(page, title) {
  try {
    await page.goto('https://www.walmart.com', { timeout: 60000 });
    await page.fill('input[aria-label="Search Walmart.com"]', title);
    await page.keyboard.press('Enter');
    await page.waitForSelector('[data-item-id]', { timeout: 60000 });
    // Walmart 价格选择器可能有变化，尝试多种方式
    const price = await page.locator('[data-automation-id="product-price"], span[class*=price]').first().textContent();
    // 点击第一个商品，进入详情页
    await page.locator('[data-item-id]').first().click();
    await page.waitForLoadState('domcontentloaded', { timeout: 60000 });
    // 抓取描述（例如，在 [data-automation-id="product-description"] 中）
    const description = await page.locator('[data-automation-id="product-description"]').textContent().catch(() => '');
    // 抓取商品尺寸（例如，在 [data-automation-id="product-specifications"] 中，或通过正则匹配）
    const sizeText = await page.locator('[data-automation-id="product-specifications"]').textContent().catch(() => '');
    const sizeMatch = sizeText.match(/size\s*:?\s*([^,;]+)/i);
    const productSize = sizeMatch ? sizeMatch[1].trim() : '';
    console.log("Walmart 抓取字段:", { price, description, productSize });
    return { price: price ? price.trim() : '', description, productSize };
  } catch (e) {
    console.log("Walmart 抓取错误:", e);
    return { price: '', description: '', productSize: '' };
  }
}

async function searchTemu(page, title) {
  try {
    await page.goto('https://www.temu.com', { timeout: 60000 });
    // 假设搜索框的 aria-label 为 "Search Temu"
    await page.fill('input[aria-label="Search Temu"]', title);
    await page.keyboard.press('Enter');
    await page.waitForSelector('[data-item-id]', { timeout: 60000 });
    const price = await page.locator('[data-automation-id="product-price"], span[class*=price]').first().textContent();
    // 点击第一个商品，进入详情页
    await page.locator('[data-item-id]').first().click();
    await page.waitForLoadState('domcontentloaded', { timeout: 60000 });
    // 抓取描述（例如，在 [data-automation-id="product-description"] 中）
    const description = await page.locator('[data-automation-id="product-description"]').textContent().catch(() => '');
    // 抓取商品尺寸（例如，在 [data-automation-id="product-specifications"] 中，或通过正则匹配）
    const sizeText = await page.locator('[data-automation-id="product-specifications"]').textContent().catch(() => '');
    const sizeMatch = sizeText.match(/size\s*:?\s*([^,;]+)/i);
    const productSize = sizeMatch ? sizeMatch[1].trim() : '';
    console.log("Temu 抓取字段:", { price, description, productSize });
    return { price: price ? price.trim() : '', description, productSize };
  } catch (e) {
    console.log("Temu 抓取错误:", e);
    return { price: '', description: '', productSize: '' };
  }
}

async function searchGoogle(page, title) {
  try {
    await page.goto('https://www.google.com', { timeout: 60000 });
    await page.fill('input[name="q"]', title + " buy");
    await page.keyboard.press('Enter');
    await page.waitForSelector('a[href]', { timeout: 60000 });
    // 过滤掉 Amazon、eBay、Walmart、Temu 的链接，抓取第一个独立站链接
    const links = await page.$$('a[href]');
    let indyLink = null;
    for (const link of links) {
      const href = await link.getAttribute('href');
      if (href && !href.includes('amazon.com') && !href.includes('ebay.com') && !href.includes('walmart.com') && !href.includes('temu.com')) {
        indyLink = href;
        break;
      }
    }
    if (indyLink) {
      await page.goto(indyLink, { timeout: 60000 });
      await page.waitForLoadState('domcontentloaded', { timeout: 60000 });
      // 抓取价格（假设独立站价格在 span[class*=price] 中）
      const price = await page.locator('span[class*=price]').first().textContent().catch(() => '');
      // 抓取描述（假设描述在 meta[name="description"] 中，或页面中某 div 中）
      const description = await page.locator('meta[name="description"]').getAttribute('content').catch(() => '');
      // 抓取商品尺寸（假设尺寸在页面某 div 中，或通过正则匹配）
      const sizeText = await page.locator('div:has-text("size")').textContent().catch(() => '');
      const sizeMatch = sizeText.match(/size\s*:?\s*([^,;]+)/i);
      const productSize = sizeMatch ? sizeMatch[1].trim() : '';
      console.log("Google 抓取字段:", { price, description, productSize });
      return { price: price ? price.trim() : '', description, productSize };
    } else {
      console.log("Google 未找到独立站链接");
      return { price: '', description: '', productSize: '' };
    }
  } catch (e) {
    console.log("Google 抓取错误:", e);
    return { price: '', description: '', productSize: '' };
  }
}

async function scrapePrices(titles) {
  const browser = await chromium.launch({ headless: false });
  const page = await browser.newPage();
  const results = {};
  // 只抓取第一个商品，方便调试
  const title = titles[0];
  console.log(`正在抓取: ${title}`);
  const amazon = await searchAmazon(page, title);
  console.log("Amazon 抓取结果:", amazon);
  const ebay = await searchEbay(page, title);
  console.log("eBay 抓取结果:", ebay);
  const walmart = await searchWalmart(page, title);
  console.log("Walmart 抓取结果:", walmart);
  const temu = await searchTemu(page, title);
  console.log("Temu 抓取结果:", temu);
  const google = await searchGoogle(page, title);
  console.log("Google 抓取结果:", google);
  results[title] = { amazon, ebay, walmart, temu, google };
  await browser.close();
  fs.writeFileSync("output.md", JSON.stringify(results, null, 2));
  console.log("抓取结果已写入 output.md");
  return results;
}

async function main() {
  const [titles, allRows] = await Promise.all([
    readProductTitles(INPUT_CSV),
    readAllRows(INPUT_CSV)
  ]);
  const priceMap = await scrapePrices(titles);
  // 构造输出
  const csvWriter = createObjectCsvWriter({
    path: OUTPUT_CSV,
    header: [
      ...Object.keys(allRows[0]),
      { id: 'Amazon Price', title: 'Amazon Price' },
      { id: 'Amazon Description', title: 'Amazon Description' },
      { id: 'Amazon Product Size', title: 'Amazon Product Size' },
      { id: 'eBay Price', title: 'eBay Price' },
      { id: 'eBay Description', title: 'eBay Description' },
      { id: 'eBay Product Size', title: 'eBay Product Size' },
      { id: 'Walmart Price', title: 'Walmart Price' },
      { id: 'Walmart Description', title: 'Walmart Description' },
      { id: 'Walmart Product Size', title: 'Walmart Product Size' },
      { id: 'Temu Price', title: 'Temu Price' },
      { id: 'Temu Description', title: 'Temu Description' },
      { id: 'Temu Product Size', title: 'Temu Product Size' },
      { id: 'Google Price', title: 'Google Price' },
      { id: 'Google Description', title: 'Google Description' },
      { id: 'Google Product Size', title: 'Google Product Size' }
    ]
  });

  const outputRows = allRows.map(row => {
    const title = row['Title'];
    const amazon = priceMap[title]?.amazon || { price: '', description: '', productSize: '' };
    const ebay = priceMap[title]?.ebay || { price: '', description: '', productSize: '' };
    const walmart = priceMap[title]?.walmart || { price: '', description: '', productSize: '' };
    const temu = priceMap[title]?.temu || { price: '', description: '', productSize: '' };
    const google = priceMap[title]?.google || { price: '', description: '', productSize: '' };
    return {
      ...row,
      'Amazon Price': amazon.price,
      'Amazon Description': amazon.description,
      'Amazon Product Size': amazon.productSize,
      'eBay Price': ebay.price,
      'eBay Description': ebay.description,
      'eBay Product Size': ebay.productSize,
      'Walmart Price': walmart.price,
      'Walmart Description': walmart.description,
      'Walmart Product Size': walmart.productSize,
      'Temu Price': temu.price,
      'Temu Description': temu.description,
      'Temu Product Size': temu.productSize,
      'Google Price': google.price,
      'Google Description': google.description,
      'Google Product Size': google.productSize
    };
  });

  await csvWriter.writeRecords(outputRows);
  console.log(`已写入 ${OUTPUT_CSV}`);
}

main(); 

File: price_scraper.py
import csv
import asyncio
from playwright.async_api import async_playwright
import pandas as pd
from datetime import datetime
import time
import random

class PriceScraper:
    def __init__(self):
        self.results = []
        self.headers = {
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36'
        }

    async def search_amazon(self, page, product_title):
        try:
            # 访问 Amazon
            await page.goto('https://www.amazon.com')
            
            # 等待搜索框加载
            await page.wait_for_selector('#twotabsearchtextbox')
            
            # 输入搜索词
            await page.fill('#twotabsearchtextbox', product_title)
            await page.press('#twotabsearchtextbox', 'Enter')
            
            # 等待搜索结果加载
            await page.wait_for_selector('[data-component-type="s-search-result"]')
            
            # 获取第一个商品的价格
            price_element = await page.query_selector('.a-price .a-offscreen')
            if price_element:
                price_text = await price_element.text_content()
                return price_text.strip()
            return "Not found"
        except Exception as e:
            print(f"Amazon search error for {product_title}: {str(e)}")
            return "Error"

    async def search_ebay(self, page, product_title):
        try:
            # 访问 eBay
            await page.goto('https://www.ebay.com')
            
            # 等待搜索框加载
            await page.wait_for_selector('#gh-ac')
            
            # 输入搜索词
            await page.fill('#gh-ac', product_title)
            await page.press('#gh-ac', 'Enter')
            
            # 等待搜索结果加载
            await page.wait_for_selector('.s-item__price')
            
            # 获取第一个商品的价格
            price_element = await page.query_selector('.s-item__price')
            if price_element:
                price_text = await price_element.text_content()
                return price_text.strip()
            return "Not found"
        except Exception as e:
            print(f"eBay search error for {product_title}: {str(e)}")
            return "Error"

    async def search_walmart(self, page, product_title):
        try:
            # 访问 Walmart
            await page.goto('https://www.walmart.com')
            
            # 等待搜索框加载
            await page.wait_for_selector('#global-search-input')
            
            # 输入搜索词
            await page.fill('#global-search-input', product_title)
            await page.press('#global-search-input', 'Enter')
            
            # 等待搜索结果加载
            await page.wait_for_selector('[data-item-id]')
            
            # 获取第一个商品的价格
            price_element = await page.query_selector('[data-automation-id="product-price"]')
            if price_element:
                price_text = await price_element.text_content()
                return price_text.strip()
            return "Not found"
        except Exception as e:
            print(f"Walmart search error for {product_title}: {str(e)}")
            return "Error"

    async def process_product(self, page, product_title):
        # 添加随机延迟，避免被反爬
        await asyncio.sleep(random.uniform(2, 5))
        
        amazon_price = await self.search_amazon(page, product_title)
        await asyncio.sleep(random.uniform(2, 5))
        
        ebay_price = await self.search_ebay(page, product_title)
        await asyncio.sleep(random.uniform(2, 5))
        
        walmart_price = await self.search_walmart(page, product_title)
        
        return {
            'Title': product_title,
            'Amazon Price': amazon_price,
            'eBay Price': ebay_price,
            'Walmart Price': walmart_price,
            'Timestamp': datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        }

    async def scrape_prices(self, input_csv):
        # 读取输入 CSV 文件
        df = pd.read_csv(input_csv)
        product_titles = df['Title'].unique().tolist()
        
        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=True)
            context = await browser.new_context(
                user_agent=self.headers['User-Agent']
            )
            page = await context.new_page()
            
            for title in product_titles:
                if pd.isna(title):
                    continue
                    
                result = await self.process_product(page, title)
                self.results.append(result)
                print(f"Processed: {title}")
                
                # 每处理 5 个商品保存一次结果
                if len(self.results) % 5 == 0:
                    self.save_results()
            
            await browser.close()
            
        # 最后保存一次结果
        self.save_results()

    def save_results(self):
        # 将结果保存为 CSV
        output_filename = f'price_results_{datetime.now().strftime("%Y%m%d_%H%M%S")}.csv'
        df = pd.DataFrame(self.results)
        df.to_csv(output_filename, index=False)
        print(f"Results saved to {output_filename}")

async def main():
    scraper = PriceScraper()
    await scraper.scrape_prices('products_export.csv')

if __name__ == "__main__":
    asyncio.run(main()) 